book/scheduling-and-threading.md

```diff
@@ -93,137 +44,130 @@ class Task:
+[^event-loop]: Event loops, which we discussed in [Chapter
+2](graphics.md#eventloop) and [Chapter
+11](visual-effects.md#sdl-creates-the-window) are one kind of task
+runner, where the tasks to run are provided by the operating system.
+
+``` {.python replace=(self)/(self%2c%20tab)}
 class TaskRunner:
     def __init__(self):
         self.tasks = []
 
     def schedule_task(self, callback):
         self.tasks.append(callback)
+```
+
+When the time comes to run a task, our task runner can just remove
+the first task from the queue and run it:
 
+``` {.python expected=False}
+class TaskRunner:
     def run(self):
         if len(self.tasks) > 0):
             task = self.tasks.pop(0)
-            task()
-```
-
-Finally we just need to modify the main event loop to run the task runner each
-time, in case there is a script task to execute:
+            task.run()
 
-``` {.python expected=False}
-if __name__ == "__main__":
-    # ...
-    while True:
-        while sdl2.SDL_PollEvent(ctypes.byref(event)) != 0:
-            # ...
-            browser.tabs[browser.active_tab].task_runner.run()
+class Tab:
+    def __init__(self):
+        self.task_runner = TaskRunner()
 ```
 
-Of course, this is all pointless at the moment, since there aren't any tasks
-yet. Let's try out this infrastructure on a simple example: evaluating scripts
-after load. Currently the browser just evaluates them right away, but instead
-let's make script evaluationt a task.
+This is a pretty simplistic of choosing which task to run next, and
+real browsers have sophisticated *schedulers* which consider things
+like XXX.
 
-It's as simple as calling `schedule_task` with a `Task` that runs the script,
-and then continuing on with the rest of the loading logic as if nothing
-was changed:
+But even this simple scheduler lets us save tasks for later and
+execute when there's time. For example, right now, when loading a web
+page, our browser will download and run all scripts before doing its
+rendering steps. That makes pages slower to load. We can fix this by
+creating tasks for running scripts later.
 
 ``` {.python expected=False}
 class Tab:
-    def __init__(self):
-        self.task_runner = TaskRunner()
-
     def run_script(self, url, body):
         try:
             print("Script returned: ", self.js.run(body))
         except dukpy.JSRuntimeError as e:
             print("Script", url, "crashed", e)
 
     def load(self):
-        # ...
-
         for script in scripts:
             # ...
             header, body = request(script_url, url)
             self.task_runner.schedule_task(
                 Task(self.run_script, script_url, body))
 ```
 
-That's it! Now our browser will not run scripts until after `load` has completed
-and the event loop comes around again. But before continuing, let's consider
-why this change is interesting---is it anything other than just some
-refactoring?
+To run those tasks, we need to call the `run` method on our
```

Comment:
I think this makes more sense above the example of loading tasks, since without it it's not clear when tasks would ever run.

Reply:
Done